src/type-extensions/promise.ts

```diff
@@ -0,0 +1,137 @@
+                            throw new Error('promise has no instance data')
+                        }
+                        let promiseResult: { status: 'fulfilled' | 'rejected'; value: any } | undefined = undefined
+
+                        const awaitPromise = self
+                            .then((res) => {
+                                promiseResult = { status: 'fulfilled', value: res }
+                            })
+                            .catch((err) => {
+                                promiseResult = { status: 'rejected', value: err }
+                            })
+
+                        const continuance = this.thread.cmodule.module.addFunction((continuanceState: LuaState) => {
+                            // If this yield has been called from within a coroutine and so manually resumed
+                            // then there may not yet be any results. In that case yield again.
+                            if (!promiseResult) {
+                                // 1 is because the initial yield pushed a promise reference so this pops
+                                // it and re-returns it.
+                                // 0 because this is called between resumes so the first one should've
+                                // popped the promise before returning the result. This is true within
+                                // Lua's coroutine.resume too.
+                                return thread.cmodule.lua_yieldk(functionThread.address, 0, 0, continuance)
+                            }
+
+                            thread.cmodule.module.removeFunction(continuance)
+                            const continuanceThread = thread.stateToThread(continuanceState)
+
+                            if (!promiseResult) {
+                                promiseResult = { status: 'rejected', value: new Error('continuance called with no result') }
+                            }
+
+                            if (promiseResult.status === 'rejected') {
+                                continuanceThread.pushValue(promiseResult.value || new Error('promise rejected with no error'))
+                                return this.thread.cmodule.lua_error(continuanceState)
+                            }
+
+                            continuanceThread.pushValue(promiseResult.value)
+                            return 1
+                        }, 'iiii')
+
+                        functionThread.pushValue(awaitPromise)
+                        return thread.cmodule.lua_yieldk(functionThread.address, 1, 0, continuance)
+                    },
+                    { receiveThread: true, rawResult: true },
+                ),
+            })
+            thread.cmodule.lua_setfield(thread.address, metatableIndex, '__index')
+        }
+        // Pop the metatable from the stack.
+        thread.cmodule.lua_pop(thread.address, 1)
+
+        // Lastly create a static Promise constructor.
+        thread.set('Promise', {
```

Comment:
So far I have been trying not to create new variables, this can be a little incisive if the user doesn't want them. Maybe it's better not to do this or leave it opt-in, so there are no surprises.

Reply:
Sure can do, will have it off by default and change openStandardLibs in engine to be an options object